testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,129 @@
+static FULLNODE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"fn-(\d+)").unwrap());
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+    prometheus: Prometheus,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster, prometheus: Prometheus) -> FullNodeHealthCheck {
+        Self {
+            cluster,
+            prometheus,
+        }
+    }
+
+    pub fn state_sync_committed_version(
+        &self,
+        start: Duration,
+        end: Duration,
+        app: &str,
+    ) -> Result<HashMap<String, f64>> {
+        libra_retrier::retry(libra_retrier::fixed_retry_strategy(1_000, 5), || {
+            self.prometheus
+                .query_range_max(
+                    format!("libra_state_sync_committed_version{{app=\"{}\"}}", app),
+                    &start,
+                    &end,
+                    10, /* step */
+                )
+                .map_err(|e| format_err!("No sync committed version data: {}", e))
+        })
+    }
+}
+
+pub fn node_index(peer_name: String) -> String {
+    if let Some(cap) = VAL_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    if let Some(cap) = FULLNODE_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    panic!(
+        "Failed to parse peer name {} into validator_index",
+        peer_name
+    )
+}
+
+pub fn convert_index(input: HashMap<String, f64>) -> HashMap<String, f64> {
+    let mut res = HashMap::new();
+    for (peer_name, version) in input {
+        let peer_name = node_index(peer_name);
+        res.insert(peer_name, version);
+    }
+    res
+}
+
+impl HealthCheck for FullNodeHealthCheck {
+    fn on_event(&mut self, _event: &ValidatorEvent, _ctx: &mut HealthCheckContext) {}
+
+    fn verify(&mut self, ctx: &mut HealthCheckContext) {
+        let validator = self.cluster.validator_instances();
+        let buffer = Duration::from_secs(30);
+        let timestamp = unix_timestamp_now() - buffer;
+        let res_fullnode = match FullNodeHealthCheck::state_sync_committed_version(
+            &self,
+            timestamp,
+            timestamp,
+            APP_LIBRA_FULLNODE,
+        ) {
+            Err(_e) => return,
+            Ok(s) => convert_index(s),
+        };
+        let res_validator = match FullNodeHealthCheck::state_sync_committed_version(
+            &self,
+            timestamp,
+            timestamp,
+            APP_LIBRA_VALIDATOR,
+        ) {
+            Err(_e) => return,
+            Ok(s) => convert_index(s),
+        };
+        for (node_index, version) in &res_fullnode {
+            if res_validator.contains_key(node_index) {
+                let validator_version = res_validator[node_index];
+                // we define tolerance of version diff is 2000
+                if (version - validator_version).abs() > 20000 as f64 {
```

Comment:
Number here is 20k, is it a typo?
Seem like a huge gap?

Reply:
oh, 20k is not a typo, comment is typo, it should be 20k

Reply:
20k is a lot, can you give a link to prometheus graph that shows that full node and validator can diverge this much in regular state?
Each version is basically a transaction, 20k transaction means full nodes is 20 seconds behind. That's a lot?

Reply:
based on my collecting datas last week, I do find same case which full node and around 16k difference with validator but test can still go through under that difference

Reply:
So the question is not whether test can go through, it's mostly what is the minimal difference that covers normal operation.

For example, can you run `--emti-tx` for 10 minutes, and post here link to prometheus for this run? We can analyze version graph during this run and see how far full node is behind, and use this as a number for theshold

Reply:
1717 |  
-- | --
5339 | TRUE
998 |  
500 |  
251 |  
9050 | TRUE
750 |  
0 |  
2287 | TRUE
1001 |  
250 |  
2579 | TRUE
5255 | TRUE
1732 |  
1507 |  
751 |  
11629 | TRUE
1084 |  
6497 | TRUE
1787 |  
1797 |  
466 |  
3745 | TRUE
5496 | TRUE
6858 | TRUE
1546 |  
1787 |  
 there are one set of test difference of normal operation, I can run --emti-tx for more data